lua_modules/websocket/websocket.lua

```diff
@@ -0,0 +1,496 @@
+-- @param wsurl A string of the websocket URL to connect like "ws://10.0.0.104:8980/ws"
+-- @param callback A function to get called back after a connection is established to the websocket URL
+-- @returns nil
+function m.connect(wsurl, callbackOnConnected)
+  
+  if wsurl ~= nil then m.wsUrl = wsurl end
+  if callbackOnConnected ~= nil then m.callbackOnConnected = callbackOnConnected end 
+  
+  -- Get host from websocket url 
+  local host, port, path = string.match(m.wsUrl, 'ws://(.-):(.-)/(.*)')
+  if host and port and path then
+    path = "/" .. path 
+  else
+    host, path = string.match(m.wsUrl, 'ws://(.-)/(.*)')
+    if host and path then
+      port = 80
+      path = "/" .. path 
+    else
+      host = string.match(m.wsUrl, 'ws://(.-)')
+      if host then
+        port = 80
+        path = "/"
+      else
+        print("Could not parse websocket URL?", m.wsUrl)
+      end
+      
+    end
+  end
+  
+  print("Websocket connecting to:", host, "port:", port, "path:", path)
+  
+  local body = "GET " .. path .. " HTTP/1.1\r\n"
+  body = body .. "Host: " .. host .. "\r\n"
+  body = body .. "Upgrade: websocket\r\n"
+  body = body .. "Connection: Upgrade\r\n"
+  body = body .. "Sec-WebSocket-Key: x3JJHMbDL1EzLkh9GBhXDw==\r\n"
+  body = body .. "Sec-WebSocket-Protocol: chat, superchat\r\n"
+  body = body .. "Sec-WebSocket-Version: 13\r\n"
+  body = body .. "Origin: esp32\r\n"
+  body = body .. "\r\n"
+  
+  local sk = net.createConnection(net.TCP)
+  
+  local buffer
+  local isHdrRecvd = false
+  sk:on("receive", function(sck, c) 
+    -- print("Got receive: " .. c)
+    
+    -- see if we are good to go here 
+    if isHdrRecvd == false then 
+      -- look for header    
+      -- look for HTTP/1.1 101
+      if string.match(c, "HTTP/1.1 101(.*)\r\n\r\n") then 
+        print("Websocket found hdr")
+        isHdrRecvd = true
+        m.isConnected = true
+        if m.onConnectedCallback ~= nil then 
+          node.task.post(node.task.LOW_PRIORITY, function()
+            m.onConnectedCallback(host, port, path) 
+          end)
+        end 
+      end 
+    else 
+      -- we can start to deliver incoming data 
+  
+      m.decodeFrame(c)
+    end 
+    
+  end)
+  sk:on("sent", function()
+    -- print("Websocket sent data")
+  end)
+  sk:on('disconnection', function(errcode)
+    print("Websocket disconnection. err:", errcode)
+    m.isConnected = false
+    if m.onDisconnectCallback then m.onDisconnectCallback() end 
+  end)
+  sk:on('reconnection', function(errcode)
+    print('Websocket reconnection. err:', errcode)
+  end)
+  sk:on("connection", function(sck, c)
+    -- print("Websocket got connection. Sending TCP msg. body:")
+    print(body)
+    -- m.sock = sck 
+    sck:send(body)
+    
+    m.pingStart()
+  end)
+  
+  m.sock = sk
+  sk:connect(port, host)
+  
+end
+
+-- currently only supports minimal frame decode
+m.lenExpected = 0
+m.buffer = ""
+-- private method
```

Comment:
Yes, I should be cleaner on that. I feel like most people ignore public/private in Lua, so I was just doing it more as a comment than a strict approach because Lua is pretty relaxed.